core/state.py

```diff
@@ -680,9 +680,13 @@ def SetVar(self, lval, value, new_flags, lookup_mode):
           strs.extend([None] * n)
           strs[lval.index] = value.s
       else:
-        # TODO:
-        # - This is a bug, because a[2]=2 creates an array of length ONE, even
-        # though the index is two.
+        # When the array doesn't exist yet, it is created filled with None.
+        # Access to the array needs to explicitly filter those sentinel values.
+        # It also wastes memory. But indexed access is fast.
+
+        # What should be optimized for? Bash uses a linked list. Random access
+        # takes linear time, but iteration skips unset entries automatically.
+
```

Comment:
I didn't want to completely remove the comment, because it might turn out that the performance of this representation isn't good enough for some use case. It really depends on whether anyone wants to set `a[1000000000]=1` and expects it to be as fast as in bash.